src/test/java/com/android/volley/toolbox/DiskBasedAsyncCacheTest.java

```diff
@@ -0,0 +1,449 @@
+    public void testManyResponseHeaders() throws ExecutionException, InterruptedException {
+        Cache.Entry entry = new Cache.Entry();
+        entry.data = new byte[0];
+        entry.responseHeaders = new HashMap<>();
+        for (int i = 0; i < 0xFFFF; i++) {
+            entry.responseHeaders.put(Integer.toString(i), "");
+        }
+        putEntry("key", entry).get();
+    }
+
+    @Test
+    public void initializeIfRootDirectoryDeleted() throws ExecutionException, InterruptedException {
+        temporaryFolder.delete();
+
+        Cache.Entry entry = randomData(101);
+        putEntry("key1", entry).get();
+
+        assertThat(getEntry("key1").get(), is(nullValue()));
+
+        // confirm that we can now store entries
+        putEntry("key2", entry).get();
+        assertThatEntriesAreEqual(getEntry("key2").get(), entry);
+    }
+
+    /* Test helpers */
+
+    /** Puts entry into the cache, and returns a CompletableFuture, to let us know when it's done */
+    private CompletableFuture<Void> putEntry(final String key, Cache.Entry entry) {
+        final CompletableFuture<Void> put = new CompletableFuture<>();
+        cache.put(
+                key,
+                entry,
+                new AsyncCache.OnCompleteCallback() {
+                    @Override
+                    public void onComplete() {
+                        put.complete(null);
+                    }
+                });
+        return put;
+    }
+
+    /** Gets an entry from the cache, and returns a CompletableFuture containing the entry */
+    private CompletableFuture<Cache.Entry> getEntry(final String key) {
+        final CompletableFuture<Cache.Entry> get = new CompletableFuture<>();
+        cache.get(
+                key,
+                new AsyncCache.OnGetCompleteCallback() {
+                    @Override
+                    public void onGetComplete(@Nullable Cache.Entry entry) {
+                        get.complete(entry);
+                    }
+                });
+        return get;
+    }
+
+    private CompletableFuture<Void> removeEntry(final String key) {
+        final CompletableFuture<Void> remove = new CompletableFuture<>();
+        cache.remove(
+                key,
+                new AsyncCache.OnCompleteCallback() {
+                    @Override
+                    public void onComplete() {
+                        remove.complete(null);
+                    }
+                });
+        return remove;
+    }
+
+    private CompletableFuture<Void> invalidateEntry(final String key, final boolean fullExpire) {
+        final CompletableFuture<Void> remove = new CompletableFuture<>();
+        cache.invalidate(
+                key,
+                fullExpire,
+                new AsyncCache.OnCompleteCallback() {
+                    @Override
+                    public void onComplete() {
+                        remove.complete(null);
+                    }
+                });
+        return remove;
+    }
+
```

Comment:
Optional: Could be worth creating a CacheTestUtil with the methods below since they are used in the DiskBasedCacheTest as well

Reply:
Currently, almost all of the statements fall on one line, and I think making this change would make a lot of them multiple line statements and harder to read. With that being said, it is a decent amount of duplicated code, so I think there are benefits to either options.

Reply:
I'd personally learn towards not having duplicate code, and hopefully the statements are still pretty readable, but it's up to you :)